runtime/sema/type.go

```diff
@@ -6317,6 +6439,17 @@ func (t *RestrictedType) RewriteWithRestrictedTypes() (Type, bool) {
 	return t, false
 }
 
+func (t *RestrictedType) Map(gauge common.MemoryGauge, typeParamMap map[*TypeParameter]*TypeParameter, f func(Type) Type) Type {
+	restrictions := make([]*InterfaceType, 0, len(t.Restrictions))
+	for _, restriction := range t.Restrictions {
+		if mappedRestriction, isRestriction := restriction.Map(gauge, typeParamMap, f).(*InterfaceType); isRestriction {
+			restrictions = append(restrictions, mappedRestriction)
+		}
```

Comment:
`RestrictedType.Restrictions` is `[]*InterfaceType`, so `restriction` is `*InterfaceType`, and therefore `Map` should always return the same type:

```suggestion
		mappedRestriction, ok := restriction.Map(gauge, typeParamMap, f).(*InterfaceType)
		if !ok {
			panic(errors.NewUnreachableError())
		}
		restrictions = append(restrictions, mappedRestriction)
```

(It would be great if we could express this in Go directly, but its generic system has no notion of a "self type")

Reply:
Ah, I see. It might be better to just panic in the case where the requirement that restrictions need to be interfaces does not hold, instead of just silently ignoring the result of the mapping function